src/lib/memory/blob_manager.cpp

```diff
@@ -2,34 +2,48 @@
 
 namespace homeobject {
 
-BlobManager::Result< blob_id > MemoryHomeObject::_put_blob(ShardInfo const& shard, Blob&& blob) {
-    auto lg = std::scoped_lock(_index_lock);
-    auto [btree_it, h1] = _in_memory_index.try_emplace(shard.id, std::make_pair(btree(), blob_id(0ull)));
-    RELEASE_ASSERT(_in_memory_index.end() != btree_it, "Could not create BTree!");
+BlobManager::Result< MemoryHomeObject::index_svc::iterator > MemoryHomeObject::_find_index(shard_id shard) {
+    auto lg = std::shared_lock(_index_lock);
+    if (auto index_it = _in_memory_index.find(shard); _in_memory_index.end() == index_it)
+        return folly::makeUnexpected(BlobError::UNKNOWN_SHARD);
+    else
+        return index_it;
+}
 
-    auto route = BlobRoute{shard.id, btree_it->second.second++};
-    LOGDEBUGMOD(homeobject, "Writing BLOB {} to: BlkId:[{}]", route.blob, fmt::ptr(&blob));
+BlobManager::Result< blob_id > MemoryHomeObject::_put_blob(ShardInfo const& shard, Blob&& blob) {
+    auto e = _find_index(shard.id);
+    if (!e) return folly::makeUnexpected(e.error());
+    auto index_it = e.value();
 
-    auto [_, happened] = btree_it->second.first.try_emplace(route, std::move(blob));
+    auto route = BlobRoute{shard.id, 0ull};
+    auto bt_lg = std::scoped_lock(index_it->second._btree_lock);
+    route.blob = index_it->second._shard_seq_num++;
+    auto [new_it, happened] = index_it->second._btree.try_emplace(route, std::move(blob));
     RELEASE_ASSERT(happened, "Generated duplicate BlobRoute!");
 
+    LOGDEBUGMOD(homeobject, "Wrote BLOB {} to: BlkId:[{}]", route.blob, fmt::ptr(&new_it->second));
     return route.blob;
 }
 
 BlobManager::Result< Blob > MemoryHomeObject::_get_blob(ShardInfo const& shard, blob_id blob) const {
+    auto index_it = _in_memory_index.cend();
+    {
```

Comment:
replace this section with _find_index?  as the `iterator` will be implicitly cast to `const iterator`

Reply:
I wanted to do that; but i can't call a non-const method from const, and making it const makes it const so i can't _put on it anymore. I know there's a trick...but it's eluding me at the moment.